### PR TITLE
do not stat file for every log() site

### PR DESCRIPTION
**Pull request by Raynos**
reviewers: @markyen @mranney

### CODE REVIEW

File: lib/mkdir-daily-rotate-file.js [hunk 1]
```diff
@@ -12,6 +14,22 @@ util.inherits(MkdirDailyRotateFile, DailyRotateFile);
 MkdirDailyRotateFile.prototype.name = 'MkdirDailyRotateFile';
 
 MkdirDailyRotateFile.prototype.open = function open(cb) {
+    var now = new Date();
+    if (!this._stream ||
+      (this.maxsize && this._size >= this.maxsize) ||
+      (this._year < now.getFullYear() || this._month < now.getMonth() || this._date < now.getDate() || this._hour < now.getHours() || this._minute < now.getMinutes())
```
Comment (markyen) [inline lib/mkdir-daily-rotate-file.js]: Is it ok that this still stats at least once every minute?

File: lib/mkdir-daily-rotate-file.js [hunk 2]
```diff
@@ -12,6 +14,22 @@ util.inherits(MkdirDailyRotateFile, DailyRotateFile);
...
+    ) {
+        // do proper stat
+        this._open(cb);
+    } else if (this._stream) {
+        return cb();
+    } else {
```
Comment (markyen) [inline lib/mkdir-daily-rotate-file.js]: This branch is unreachable

File: lib/mkdir-daily-rotate-file.js (hunk 1 above)
Comment (Raynos) [inline lib/mkdir-daily-rotate-file.js]: That's what the upstream logic does.

This is far better then doing it for every single log() callsite.

File: lib/mkdir-daily-rotate-file.js (hunk 2 above)
Comment (Raynos) [inline lib/mkdir-daily-rotate-file.js]: good point.

**Comment by markyen** [comment]
LGTM, though I wonder if it would be better to fork/contribute to DailyRotateFile rather than coupling ourselves so super tightly to its internal implementation

**Comment by Raynos** [comment]
We have a fork of `uber-winston` that we could fix. Contributing back is hard because we would have to upgrade.

**Pull request merged**